File: src/bmsx_cpp/core/host_frame.cpp
#include "core/console.h"

#include "core/host_overlay_menu.h"
#include "common/time.h"
#include "input/manager.h"
#include "machine/runtime/runtime.h"

#include <chrono>

namespace bmsx {
namespace {
constexpr double MAX_FRAME_DELTA_MS = 250.0;
}

void ConsoleCore::runHostFrame(
	Runtime& runtime,
	MicrotaskQueue& microtasks,
	f64 deltaTime,
	bool platformPaused
) {
	if (!acceptHostFrame(deltaTime)) {
		return;
	}
	try {
		const auto tickStart = std::chrono::steady_clock::now();
		runtime.screen.recordHostFrame();

		double hostDeltaMs = deltaTime * 1000.0;
		if (hostDeltaMs > MAX_FRAME_DELTA_MS) {
			hostDeltaMs = MAX_FRAME_DELTA_MS;
		}
		const double hostDeltaSeconds = hostDeltaMs / 1000.0;
		m_delta_time = hostDeltaSeconds;
		m_total_time += hostDeltaSeconds;
		m_frame_count += 1;
		runtime.frameLoop.currentTimeSeconds = m_total_time;
		m_fps = 1.0 / hostDeltaSeconds;

		Input::instance().pollInput();
		const bool hostMenuActive = hostOverlayMenu().tickInput(*this);

		runtime.screen.clearPresentation();
		if (!platformPaused && !hostMenuActive) {
			const i64 previousTickSequence = runtime.frameScheduler.lastTickSequence;
			m_delta_time = runtime.timing.frameDurationMs / 1000.0;
			// Handle program reload request at the frame boundary (TS parity: no ConsoleCore in CartBootState)
			if (runtime.isRebootRequested()) {
				runtime.clearRebootRequest();
				runtime.frameScheduler.clearQueuedTime();
				if (!rebootLoadedRom()) {
					runtime.handleLuaError("Runtime fault: reboot to bootrom failed.");
					return;
				}
			}
			runtime.frameScheduler.run(runtime, hostDeltaMs);
			runtime.screen.syncAfterRuntimeUpdate(runtime, previousTickSequence);
		} else {
			runtime.frameScheduler.clearQueuedTime();
		}
		m_delta_time = hostDeltaSeconds;

		microtasks.flush();

		m_last_tick_timing.totalMs = to_ms(std::chrono::steady_clock::now() - tickStart);

		if (hostMenuActive && m_view) {
			hostOverlayMenu().queueRenderCommands(*this, *m_view);
		} else {
			const bool hostOverlayQueued = m_view && hostOverlayMenu().queueFrameOverlayCommands(*this, *m_view);
			if (hostOverlayQueued) {
				runtime.screen.requestHeldPresentation();
			}
		}
		if (hostMenuActive) {
			runtime.screen.requestHeldPresentation();
		}
		runtime.screen.render(*this, runtime, platformPaused);
	} catch (const std::exception& e) {
		runtime.frameLoop.abandonFrameState(runtime);
		runtime.handleLuaError(e.what());
	} catch (...) {
		runtime.frameLoop.abandonFrameState(runtime);
		runtime.handleLuaError("Unhandled host frame exception.");
	}
}

} // namespace bmsx

File: src/bmsx/core/host_frame.ts
import { consoleCore } from './console';
import { hostOverlayMenu } from './host_overlay_menu';
import * as workbenchMode from '../ide/workbench/mode';
import type { Runtime } from '../machine/runtime/runtime';
import { createRuntimeFrameStepResult, runRuntimeFrameStepInto } from '../machine/runtime/frame/step';

const MAX_HOST_FRAME_DELTA_MS = 250;
const hostFrameStepResult = createRuntimeFrameStepResult();

export function runConsoleHostFrame(runtime: Runtime, currentTime: number, runReady: boolean): void {
	const console = consoleCore;
	if (!console.running) {
		return;
	}
	const screen = runtime.screen;
	let hostDeltaMs = 0;
	try {
		console.input.pollInput();
		screen.beginHostFrame(currentTime);
		workbenchMode.tickIdeInput(runtime);
		workbenchMode.tickTerminalInput(runtime);
		hostDeltaMs = Math.min(currentTime - runtime.frameLoop.currentTimeMs, MAX_HOST_FRAME_DELTA_MS);
		runtime.frameLoop.currentTimeMs = currentTime;
		console.host_fps = 1000 / hostDeltaMs;
		const hostMenuActive = hostOverlayMenu.tickInput();

		if (hostMenuActive) {
			screen.clearPresentation();
			runtime.frameScheduler.clearQueuedTime();
			hostOverlayMenu.queueRenderCommands();
			screen.requestHeldPresentation();
			screen.presentPending(hostDeltaMs);
		} else if (console.paused) {
			hostOverlayMenu.queueFrameOverlayCommands();
			screen.presentPausedFrame(hostDeltaMs);
		} else {
			const hostOverlayQueued = hostOverlayMenu.queueFrameOverlayCommands();
			screen.clearPresentation();
			if (runtime.executionOverlayActive) {
				screen.runOverlay();
			} else if (!runReady) {
				runtime.frameScheduler.clearQueuedTime();
			} else {
				console.deltatime = runtime.timing.frameDurationMs;
				runRuntimeFrameStepInto(hostFrameStepResult, runtime, hostDeltaMs);
				screen.syncAfterRuntimeUpdate(hostFrameStepResult.previousTickSequence);
			}
			if (hostOverlayQueued) {
				screen.requestHeldPresentation();
			}
			screen.presentPending(hostDeltaMs);
		}
	} catch (error) {
		workbenchMode.surfaceHostFrameError(runtime, error, hostDeltaMs);
	}
	screen.flushDebugReport(currentTime, runtime);
}
